covid_moonshot/analysis/diffnet.py

```diff
@@ -0,0 +1,151 @@
+from arsenic import stats
+import networkx as nx
+import numpy as np
+import logging
+
+_logger = logging.getLogger("DiffNet")
+
+
+EXPERIMENTAL_ERROR = 0.5 ## pIC50
+
+def _ic50_to_dG(ic50, s_conc=375E-9, Km=40E-6):
+    """Converts IC50 (in M units) to DG
+
+    Parameters
+    ----------
+    ic50 : float
+        IC50 in M
+    s_conc : float, default=375E-9
+        Substrate concentration in M
+    Km : float, default=40E-6
+        Substrate concentration for half-maximal enzyme activity
+
+    Returns
+    -------
+    type
+        Description of returned object.
+
+    """
+    if ic50 > 1E-5:
+        _logger.warning('Expecting IC50 in M units. Please check.')
+    Ki = ic50 / (1 + (s_conc/Km))
+    return np.log(Ki)
+
+
+def separate_graphs(graphs):
+    graphs_to_replace = []
+    new_graphs = []
+    # now need to check that all of the graphs are weakly connected
+    # and if not, split them up
+    for name, g in graphs.items():
+        if not nx.is_weakly_connected(g):
+            graphs_to_replace.append(name)
+            for j, subgraph in enumerate(nx.weakly_connected_components(g)):
+                # this should perseve all necessary nodes/edges
+                new_graphs[name+f'_{j}'] = reduce_graph(g, subgraph)
+    # now get rid of non-connected graphs
+    for name in graphs_to_replace:
+        del graphs[name]
+
+    # and add the new graphs
+    graphs.update(new_graphs)
+
+    return graphs
+
+
+def reduce_graph(g, nodes_to_keep):
+    """ Takes a big graph and retains only required nodes
+    This can be useful for minimising large graphs,
+    or breaking down unconnected graphs
+
+    Any information of the graph, nodes (to be kept), and edges (where both nodes are to be kept) are retained
+
+    Parameters
+    ----------
+    g : nx.DiGraph
+        Large graph to simplify
+    nodes_to_keep : list
+        List of nodes that should be retained
+
+    Returns
+    -------
+    nx.DiGraph
+        Smaller subset graph of g
+
+    """
+    import copy
+    new_g = copy.deepcopy(g)
+    new_g.remove_nodes_from([i for i in new_g.nodes()
+                             if i not in nodes_to_keep])
+    return new_g
+
+
+def combine_relative_free_energies(results):
```

Comment:
I think we'll eventually want to transition this to taking a `List[Run]` or `Analysis` object rather than a dictionary (but this can be done later)

Reply:
(I think we will want to merge #65 before doing this anyway)